### PR TITLE
types: fix #4954, insert column char(4) with latin1 charset by incorrect padding

### LINKED ISSUE #4954: latin1 character set select bug?
MySQL:

```
mysql> CREATE TABLE t2 (a CHAR(5) CHARACTER SET latin1);
Query OK, 0 rows affected (0.01 sec)

mysql> INSERT INTO t2 VALUES ('oe');
Query OK, 1 row affected (0.01 sec)

mysql> INSERT INTO t2 VALUES (0xf6);
Query OK, 1 row affected (0.00 sec)

mysql> SELECT * FROM t2 WHERE a= 'oe';
+------+
| a    |
+------+
| oe   |
+------+
1 row in set (0.00 sec)

mysql> SELECT * FROM t2 WHERE a= 0xf6;
+------+
| a    |
+------+
| ö    |
+------+
1 row in set (0.00 sec)
```

TiDB:

```
mysql> CREATE TABLE t2 (a CHAR(5) CHARACTER SET latin1);
Query OK, 0 rows affected (0.08 sec)

mysql> INSERT INTO t2 VALUES ('oe');
Query OK, 1 row affected (0.01 sec)

mysql> INSERT INTO t2 VALUES (0xf6);
Query OK, 1 row affected (0.00 sec)

mysql> SELECT * FROM t2 WHERE a= 'oe';
Empty set (0.00 sec)

mysql> SELECT * FROM t2 WHERE a= 0xf6;
Empty set (0.00 sec)

mysql> select tidb_version();
+-----------------------------------------------------------------------------------------------------------------------------------------+
| tidb_version()                                                                                                                          |
+-----------------------------------------------------------------------------------------------------------------------------------------+
| Release Version: 1.0.0
Git Commit Hash: 460e1b8484a9834c46b6344ddd489c9ce4bc7e65
Git Branch: master
UTC Build Time: 2017-10-26 07:50:15 |
+-----------------------------------------------------------------------------------------------------------------------------------------+
```

### PR DESCRIPTION
**Pull request by winkyao**
to #4954

### CODE REVIEW

**Comment by winkyao** [comment]
@XuHuaiyu @coocood PTAL

File: plan/stats.go [hunk 1]
```diff
@@ -207,6 +207,7 @@ func (p *LogicalAggregation) prepareStatsProfile() *statsProfile {
 	return p.profile
 }
 
+// prepareStatsProfile prepare stats profile.
```
Comment (zimulala) [inline plan/stats.go]: s/prepare/prepares

File: plan/new_physical_plan_builder.go [hunk 2]
```diff
@@ -187,6 +187,7 @@ func (p *LogicalJoin) getIndexJoinByOuterIdx(outerIdx int) []PhysicalPlan {
 	return nil
 }
 
+// getChildrenPossibleProps get children possible props.:
```
Comment (zimulala) [inline plan/new_physical_plan_builder.go]: s/get/gets

File: store/tikv/lock_resolver.go [hunk 3]
```diff
@@ -33,7 +33,7 @@ type LockResolver struct {
 	store Storage
 	mu    struct {
 		sync.RWMutex
-		// Cache resolved txns (FIFO, txn id -> txnStatus).
+		// resolved Cache txns (FIFO, txn id -> txnStatus).
```
Comment (zimulala) [inline store/tikv/lock_resolver.go]: s/Cache/caches

File: expression/builtin_compare.go [hunk 4]
```diff
@@ -744,7 +750,7 @@ func (b *builtinIntervalIntSig) evalInt(row types.Row) (int64, bool, error) {
 	return int64(idx), err != nil, errors.Trace(err)
 }
 
-// All arguments are treated as integers.
+// binSearch All arguments are treated as integers.
```
Comment (zimulala) [inline expression/builtin_compare.go]: Update this comments.

File: store/tikv/client.go [hunk 5]
```diff
@@ -113,6 +113,7 @@ func (a *connArray) Close() {
 	}
 }
 
+// rpcClient is rpc client struct
```
Comment (zimulala) [inline store/tikv/client.go]: s/rpc/RPC
Add .

**Comment by zimulala** [comment]
LGTM

**Comment by XuHuaiyu** [comment]
/run-all-test
LGTM

File: util/types/datum.go [hunk 6]
```diff
@@ -791,18 +791,18 @@ func ProduceStrWithSpecifiedTp(s string, tp *FieldType, sc *variable.StatementCo
...
 				for i := range s {
 					if runeCount == flen {
-						// We don't break here because we need to iterate to the end to get runeCount.
 						truncateLen = i
+						break
 					}
 					runeCount++
 				}
-				err = ErrDataTooLong.Gen("Data Too Long, field len %d, data len %d", flen, runeCount)
+				err = ErrDataTooLong.Gen("Data Too Long, field len %d, data len %d", flen, len([]rune(s)))
```
Comment (coocood) [inline util/types/datum.go]: Why change here?

Comment (winkyao) [reply util/types/datum.go]: runeCount == len([]rune(s)), not need to iterator to count it.

Comment (coocood) [reply util/types/datum.go]: But cast string to rune slice is more expensive.

**Comment by winkyao** [comment]
@coocood PTAL

File: store/tikv/lock_resolver.go (hunk 3 above)
Comment (coocood) [inline store/tikv/lock_resolver.go]: // resolved caches resolved txns (FIFO, txn id -> txnStatus).

**Pull request merged**